packages/integration-karma/test/shadow-dom/event-in-shadow-tree/propagation.spec.js

```diff
@@ -0,0 +1,305 @@
+// Inspired from WPT:
+// https://github.com/web-platform-tests/wpt/blob/master/shadow-dom/event-inside-shadow-tree.html
+
+import { createElement } from 'lwc';
+import { extractDataIds } from 'test-utils';
+
+import Container from 'x/container';
+
+function dispatchEventWithLog(target, nodes, event) {
+    const log = [];
+
+    // We take this appraoch instead of traversing up the DOM from the target, so that we can add
+    // listeners to targets that cannot be reached by parentNode/host traversal (e.g., shadow roots
+    // of components containing assigned slots).
+    [...Object.values(nodes), document.body, document.documentElement, document].forEach((node) => {
+        node.addEventListener(
+            event.type,
+            function (event) {
+                log.push([this, event.target, event.composedPath()]);
+            }.bind(node)
+        );
+    });
+
+    target.dispatchEvent(event);
+    return log;
+}
+
+function createTestElement(parentNode) {
+    const elm = createElement('x-container', { is: Container });
+    elm.setAttribute('data-id', 'x-container');
+    parentNode.appendChild(elm);
+    return extractDataIds(elm);
+}
```

Comment:
If the element is always injected into the document body, no need to make it configurable.

```suggestion
function createTestElement() {
    const elm = createElement('x-container', { is: Container });
    elm.setAttribute('data-id', 'x-container');
    parentNode.appendChild(document.body);
    return extractDataIds(elm);
}
```